Mathlib/Computability/NFA.lean

```diff
@@ -112,6 +121,299 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    rcases eval with ⟨r, mem, step⟩
+    refine ⟨r, mem, ?_⟩
+    rcases r with _ | _
+    · exact ⟨h, p, accept, step⟩
+    · cases step
+  · rename_i _ _ ih
+    simp only [eval_append_singleton, evalFrom_append_singleton, mem_stepSet]
+    rintro ⟨p, mem, step⟩
+    exact ⟨.inr p, ih p mem, step⟩
+
+@[simp]
+theorem hmul_correct (P₁ : NFA α σ) (P₂ : NFA α σ') :
+    (P₁ * P₂).accepts = P₁.accepts * P₂.accepts := by
+  ext x
+  constructor
+  · rintro ⟨q | q, accept, eval⟩
+    · rcases accept with ⟨accept, niltail⟩
+      refine ⟨x, ⟨q, accept, ?_⟩, [], niltail, x.append_nil⟩
+      clear accept
+      induction x using List.reverseRecOn generalizing q
+      · exact eval
+      rename_i _ _ ih
+      rw [eval_append_singleton, mem_stepSet] at eval ⊢
+      rcases eval with ⟨p | p, mem, step⟩
+      · exact ⟨p, ih p mem, step⟩
+      · cases step
+    · suffices ∃ y z, y ∈ P₁.accepts ∧ q ∈ P₂.eval z ∧ y ++ z = x by
+        rcases this with ⟨y, z, y_accepts, z_eval, append⟩
+        exact ⟨y, y_accepts, z, ⟨q, accept, z_eval⟩, append⟩
+      clear accept
+      induction x using List.reverseRecOn generalizing q
+      · rcases eval with ⟨start, niltail⟩; exact ⟨[], [], niltail, start, rfl⟩
+      rename_i as a ih
+      rw [eval_append_singleton, mem_stepSet] at eval
+      rcases eval with ⟨p | p, mem, step⟩
+      · rcases step with ⟨start, r, accept, step⟩
+        refine ⟨as ++ [a], [], ⟨r, accept, ?_⟩, start, (as ++ [a]).append_nil⟩
+        rw [eval_append_singleton, mem_stepSet]
+        rw [← hmul_eval₁ p] at mem
+        exact ⟨p, mem, step⟩
+      · rcases ih p mem with ⟨y, z, accepts, eval, rfl⟩
+        refine ⟨y, z ++ [a], accepts, ?_, (y.append_assoc _ _).symm⟩
+        rw [eval_append_singleton, mem_stepSet]
+        exact ⟨p, eval, step⟩
+  · rintro ⟨y, hy, z, hz, rfl⟩
+    rcases z.eq_nil_or_concat with rfl | ⟨bs, b, rfl⟩
+    · rcases hy with ⟨q, accept, eval⟩
+      simp only [y.append_nil]
+      rw [hmul_eval₁ q] at eval
+      exact ⟨.inl q, ⟨accept, hz⟩, eval⟩
+    · rcases hz with ⟨q, accept, eval⟩
+      refine ⟨.inr q, accept, ?_⟩
+      simp only [List.concat_eq_append, ← y.append_assoc] at eval ⊢
+      rw [eval_append_singleton, mem_stepSet] at eval ⊢
+      rcases eval with ⟨p, mem, step⟩
+      refine ⟨.inr p, ?_, step⟩
+      rw [eval, evalFrom_append]
+      exact hmul_eval₂ p hy mem
+
+/-- Kleene-starring adds a state and there is no heterogeneous `KStar` class, so we fall back to
+a function definition. -/
+def kstar (P : NFA α σ) : (NFA α (Option σ)) :=
+  ⟨fun p c q ↦
+    match (p, q) with
+    | (some p, some q) => P.step p c q ∨ ∃ r, P.accept r ∧ P.step p c r ∧ P.start q
+    | _ => False,
+    Option.rec True P.start,
+    Option.rec True P.accept⟩
+
+/-- All start states of an unstarred NFA can be reached in a starred NFA by a string in the
+original's language. -/
+lemma kstar_eval_aux {P : NFA α σ} {y} (h : y ∈ P.accepts) :
```

Comment:
Thanks, I didn't know about the `private` keyword and thought `lemma` would be enough to signify the intention of being only immediately useful. I shall add the keyword, then. The other `lemma`ta might come in useful in other places though, since they describe interactions with `eval` somewhat more completely. Or would you prefer them being made private, as well?

Reply:
I think the other ones shouldn't be private